arroyo-api/src/udfs.rs

```diff
@@ -0,0 +1,248 @@
+                description: &req.description.unwrap_or_default(),
+            },
+        )
+        .await
+        .map_err(log_and_map)?;
+
+    Ok(Json(
+        api_queries::get_udf()
+            .params(
+                &client,
+                &GetUdfParams {
+                    organization_id: &auth_data.organization_id,
+                    pub_id: &pub_id,
+                },
+            )
+            .one()
+            .await
+            .map_err(log_and_map)?
+            .into(),
+    ))
+}
+
+/// Get Global UDFs
+#[utoipa::path(
+    get,
+    path = "/v1/udfs",
+    tag = "udfs",
+    responses(
+        (status = 200, description = "List of UDFs", body = GlobalUdfCollection),
+    ),
+)]
+pub async fn get_udfs(
+    State(state): State<AppState>,
+    bearer_auth: BearerAuth,
+) -> Result<Json<GlobalUdfCollection>, ErrorResp> {
+    let client = client(&state.pool).await.unwrap();
+    let auth_data = authenticate(&state.pool, bearer_auth).await.unwrap();
+
+    let udfs = api_queries::get_udfs()
+        .bind(&client, &auth_data.organization_id)
+        .all()
+        .await
+        .map_err(log_and_map)?;
+
+    Ok(Json(GlobalUdfCollection {
+        data: udfs.into_iter().map(|u| u.into()).collect(),
+    }))
+}
+
+/// Delete UDF
+#[utoipa::path(
+    delete,
+    path = "/v1/udfs/{id}",
+    tag = "udfs",
+    params(
+        ("id" = String, Path, description = "UDF id")
+    ),
+    responses(
+        (status = 200, description = "Deleted UDF"),
+    ),
+)]
+pub async fn delete_udf(
+    State(state): State<AppState>,
+    bearer_auth: BearerAuth,
+    Path(udf_pub_id): Path<String>,
+) -> Result<(), ErrorResp> {
+    let client = client(&state.pool).await.unwrap();
+    let auth_data = authenticate(&state.pool, bearer_auth).await.unwrap();
+
+    let count = api_queries::delete_udf()
+        .params(
+            &client,
+            &DeleteUdfParams {
+                organization_id: &auth_data.organization_id,
+                pub_id: &udf_pub_id,
+            },
+        )
+        .await
+        .map_err(log_and_map)?;
+
+    if count != 1 {
+        return Err(not_found("UDF"));
+    }
+
+    Ok(())
+}
+
+/// Validate UDFs
+#[utoipa::path(
+    post,
+    path = "/v1/udfs/validate",
+    tag = "udfs",
+    request_body = ValidateUdfPost,
+    responses(
+        (status = 200, description = "Validated query", body = UdfValidationResult),
+    ),
+)]
+pub async fn validate_udf(
+    State(state): State<AppState>,
+    bearer_auth: BearerAuth,
+    WithRejection(Json(validate_udf_post), _): WithRejection<Json<ValidateUdfPost>, ApiError>,
+) -> Result<Json<UdfValidationResult>, ErrorResp> {
+    let auth_data = authenticate(&state.pool, bearer_auth).await?;
+
+    // Return an ok (valid) if the controller is not available or if it fails to validate the UDFs
+
+    let mut controller = match ControllerGrpcClient::connect(state.controller_addr.clone()).await {
```

Comment:
not necessary now, but eventually we should store the connection so that we're not re-establishing it on every request